ai.py

```diff
@@ -1,98 +1,145 @@
-
-                        if board.can_move_space(dst):
-                            moves.append(Move(src, dst))
-                        elif board.can_capture_space(dst, player_sign):
-                            moves.append(Move(src, dst))
-                            break
-                        else:
-                            break
-            elif piece == 4 or piece == 5:
-                for dir in [
-                    Position(1, 0), Position(-1, 0),
-                    Position(0, 1), Position(0, -1)
-                ]:
-                    for dist in range(1, 8):
-                        dst = src + dist * dir
-
-                        if board.can_move_space(dst):
-                            moves.append(Move(src, dst))
-                        elif board.can_capture_space(dst, player_sign):
-                            moves.append(Move(src, dst))
-                            break
-                        else:
-                            break
-            elif piece == 6:
-                for dir in [
-                    Position(1, 0), Position(-1, 0),
-                    Position(0, 1), Position(0, -1),
-                    Position(1, 1), Position(-1, 1),
-                    Position(1, -1), Position(-1, -1)
-                ]:
-                    dst = src + dir
+                        break
+                    else:
+                        break
+        
+        if piece == 4 or piece == 5:
+            for dir in file_directions:
+                for dist in range(1, 8):
+                    dst = src + dist * dir
+
                     if board.can_move_space(dst):
                         moves.append(Move(src, dst))
                     elif board.can_capture_space(dst, player_sign):
                         moves.append(Move(src, dst))
-            else:
-                raise Exception("Piece {piece} not found")
-    return moves
+                        break
+                    else:
+                        break
+        elif piece == 6:
+            king_position = src
+            for dir in diagonal_directions + file_directions:
+                dst = src + dir
+                if board.can_move_space(dst) or board.can_capture_space(dst, player_sign):
+                    replacement_piece = board.move(src, dst)
+                    attacking_moves = find_attacking_moves(board, dst, -1 * player_sign)
+                    board.move(dst, src, replacement_piece)
+                    if len(attacking_moves) == 0:
+                        moves.append(Move(src, dst))
+    
+    if king_position is None:
+        return moves
+
+    valid_moves = []
+    for move in moves:
+        if move.src == king_position:
+            valid_moves.append(move)
+        else:
+            replacement_piece = board.move(move.src, move.dst)
+            attacking_moves = find_attacking_moves(board, king_position, -1 * player_sign)
+            board.move(move.dst, move.src, replacement_piece)
```

Comment:
Make sure that king is not put in check by move (could merge the above logic into this as long as the king-moved edge case is handled)